### PR TITLE
Cubic bspline hessian matrix & resolution update

### PR DESCRIPTION
**Pull request by ziyi-zhang**
Simple implementation of `eval_cubic_hessian()`.
Compatible with Nicolas's changes about wrapping mode and others.

- It returns **both positional "gradient" and "hessian"**. 
They have quite some common computation. But consequently there is no interface to compute only the "hessian".
The return type is a ~`std::tuple`~`std::pair`.

- Now all the gradients/hessian return value has **the volume resolution** multiplied.
Prior to this PR, we assume the volume has unit size (because the query point is in [0, 1]^d). I need to manually multiply the shape with the returned "gradient" in my volume optimization.
But this becomes too complicated for "hessian" matrix, where each cell needs a different multiplier. So I moved this "unit size -> volume shape transformation" jacobian inside.
This change applies to the explicit gradient computation, explicit hessian computation, and implicit AD gradient computation.

### CODE REVIEW

File: include/enoki/texture.h [hunk 1]
```diff
@@ -648,9 +651,186 @@ template <typename Value, size_t Dimension> class Texture {
         #undef EK_TEX_CUBIC_GATHER
         #undef EK_TEX_CUBIC_ACCUM
 
+        // transform volume from unit size to its resolution
+        PosF resf = PosF(m_shape_opaque);
```
Comment (wjakob) [inline include/enoki/texture.h:655]: You can define this once above and reference it. For example, this variable is used to compute `pos_f` at the beginning of the function.

File: include/enoki/texture.h [hunk 2]
```diff
@@ -11,6 +11,7 @@
 */
 
 #include <array>
+#include <tuple>
```
Comment (wjakob) [inline include/enoki/texture.h:14]: It may be non-obvious, but `<tuple>` is one of those C++ headers that we try to avoid if at all possible (the implementation of this type is surprisingly large/complex).

What you need below is simply a `std::pair<>`, which is much lighter.

File: include/enoki/texture.h [hunk 3]
```diff
@@ -356,11 +357,7 @@ template <typename Value, size_t Dimension> class Texture {
         using InterpIdx = uint32_array_t<InterpOffset>;
 
         PosF pos_(pos);
-        // This multiplication should not be recorded in the AD graph
         PosF pos_f = fmadd(pos_, PosF(m_shape_opaque), -.5f);
-        if constexpr (IsDiff)
```
Comment (wjakob) [inline include/enoki/texture.h:361]: I'm curious why this change was needed?

File: include/enoki/texture.h [hunk 4]
```diff
@@ -546,8 +542,15 @@ template <typename Value, size_t Dimension> class Texture {
...
 
-    /// Evaluate the positional gradient of a clamped cubic B-Spline from the
-    /// explicit differentiated basis functions
+    /**
+     * \brief Evaluate the positional gradient of a cubic B-Spline
+     *
+     * This implementation computes the result directly from explicit
+     * differentiated basis functions. It has no autodiff support.
+     * 
+     * The volume is assumed to have unit size. Otherwise, the resolution
```
Comment (wjakob) [inline include/enoki/texture.h:551]: `Otherwise, the resulting gradient should be multiplied by the spatial extents.`

**Review by wjakob — COMMENTED** [review]
Looks great @ziyi-zhang !

File: tests/texture.cpp [hunk 5]
```diff
@@ -394,13 +394,52 @@ ENOKI_TEST(test11_cubic_grad_pos) {
...
     assert(ek::allclose(grad_64[0][0], ref_grad[0], 1e-5f, 1e-5f));
     assert(ek::allclose(grad_64[1][0], ref_grad[1], 1e-5f, 1e-5f));
     assert(ek::allclose(grad_64[2][0], ref_grad[2], 1e-5f, 1e-5f));
     assert(ek::allclose(grad_ad, ref_grad, 1e-5f, 1e-5f));
 }
 
-ENOKI_TEST(test12_move_assignment) {
+ENOKI_TEST(test12_cubic_hessian_pos) {
+    using TensorXf = Tensor<Float>;
+    using Uint32 = ek::uint32_array_t<Float>;
```
Comment (wjakob) [inline tests/texture.cpp:406]: We usually name this type `UInt32` (capital `I`)

File: include/enoki/texture.h (hunk 3 above)
Comment (ziyi-zhang) [reply include/enoki/texture.h:361]: This function is used to replace the AD graph of positional gradients. Now we desire the multiplication with `PosF(m_shape_opaque)` to be recorded. 
The change also applies to `eval_cubic_grad()` and `eval_cubic_hessian()`.

Why make this change (result mult with shape)?
(1) If we use the grad/hessian for volume optimization, this step is always needed. I don't know any use case of grad/hessian where this multiplication is not desired.
(2) for the gradient, one can simply do `grad * shape`. But for hessian, each cell in the matrix has a different multiplier, so it might be more convenient to move this step inside the function call.

File: include/enoki/texture.h [hunk 6]
```diff
@@ -11,6 +11,7 @@
 */
 
 #include <array>
+#include <utility>
```
Comment (Speierers) [inline include/enoki/texture.h:14]: What is this needed for?

**Review by Speierers — COMMENTED** [review]
Looking great to me!

File: include/enoki/texture.h (hunk 6 above)
Comment (ziyi-zhang) [reply include/enoki/texture.h:14]: std::pair
somehow it's defined in `utility`

**Review by njroussel — COMMENTED** [review]
Looks good to me :100: !

**Comment by Speierers** [comment]
Thanks @ziyi-zhang
